File: src/main/java/org/nem/ast/AstComputer.java
package org.nem.ast;

import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.Lexer;
import org.antlr.v4.runtime.TokenStream;
import org.nem.grammar.NEMLexer;
import org.nem.grammar.NEMParser;
import org.nem.ast.visitor.FileContentVisitor;

import java.io.IOException;
import java.io.InputStream;

public class AstComputer {

	public static FileContent parseFileContent(InputStream inputStream) throws IOException {
		Lexer lexer = new NEMLexer(CharStreams.fromStream(inputStream));
		TokenStream tokenStream = new CommonTokenStream(lexer);
		NEMParser parser = new NEMParser(tokenStream);

		return parser.fileContent().accept(new FileContentVisitor());
	}
}

File: src/antlr/AstComputer.cpp

#include <antlr4-runtime.h>

#include <NEMLexer.h>
#include <NEMParser.h>

#include <nem/antlr/AstComputer.hpp>
#include <nem/antlr/ParseTreeValue.hpp>
#include <nem/antlr/ParseTreeVisitor.hpp>

namespace nem::antlr
{

ast::FileNode computeAstFile(std::istream& inStream)
{
	antlr4::ANTLRInputStream  input(inStream);
	NEMLexer				  lexer(&input);
	antlr4::CommonTokenStream tokens(&lexer);
	NEMParser				  parser(&tokens);

	ParseTreeVisitor visitor;

	auto anyValue		= parser.file()->accept(&visitor);
	auto parseTreeValue = std::any_cast<ParseTreeValue>(anyValue);
	return std::get<ast::FileNode>(parseTreeValue);
}

} // namespace nem::antlr
